doc/sphinx/language/core/basic.rst

```diff
@@ -0,0 +1,509 @@
+rest of the |Coq| manual: :term:`terms <term>` and :term:`types
+<type>` on the one hand, :term:`commands <command>` and :term:`tactics
+<tactic>` on the other hand.
+
+.. glossary::
+
+   term
+
+     The base expressions of the |Coq| language are terms.  Terms can
+     represent computable programs, program types, mathematical
+     propositions and mathematical proofs.
+
+     The syntax of terms is as follows.  Each one of the listed
+     constructs is presented in a dedicated section.  Some of these
+     constructs (like :n:`@term_product_or_fun`) are part of the core
+     language that the kernel of Coq understands and are therefore
+     described in :ref:`the present chapter <core-language>`, while
+     some others (like :n:`@term_if`) are language extensions that are
+     presented in :ref:`the next chapter <extensions>`.
+
+     .. insertprodn term qualid_annotated
+
+     .. prodn::
+        term ::= @term_product_or_fun
+        | @term_let
+        | @term_if
+        | @term_fix
+        | @term_cofix
+        | @term100
+        term100 ::= @term_cast
+        | @term10
+        term10 ::= @term_application
+        | @one_term
+        one_term ::= @ @qualid_annotated
+        | @term1
+        term1 ::= @term_projection
+        | @term_scope
+        | @term0
+        term0 ::= @qualid_annotated
+        | @sort
+        | @primitive_notations
+        | @term_evar
+        | @term_match
+        | @term_record
+        | @term_generalizing
+        | @term_ltac
+        | ( @term )
+        qualid_annotated ::= @qualid {? @univ_annot }
+
+     .. note::
+
+        Many :term:`commands <command>` and :term:`tactics <tactic>`
+        use :n:`@one_term` (in the syntax of their arguments) rather
+        than :n:`@term`.  The former need to be enclosed in
+        parentheses unless they're very simple, such as a single
+        identifier.  This avoids confusing a space-separated list of
+        terms or identifiers with a :n:`@term_application`.
+
+   type
+
+     To be valid and accepted by the |Coq| kernel, a term needs an
+     associated type.  We express this relationship by “:math:`x` *of
+     type* :math:`T`”, which we write as “:math:`x:T`”.  Informally,
+     “:math:`x:T`” can be thought as “:math:`x` *belongs to*
+     :math:`T`”.
+
+     A specificity of the |Coq| language is that types can depend on
+     terms (we say that the language is `dependently-typed
+     <https://en.wikipedia.org/wiki/Dependent_type>`_).  Because of
+     this, types and terms share a common syntax, all types are terms,
+     but not all terms are types:[#types]_
+
+     .. insertprodn type type
+
+     .. prodn::
+        type ::= @term
+
+     The kernel of |Coq| is a type checker that verifies that a term has
+     the expected type by applying a set of typing rules (see
+     :ref:`Typing-rules`).  If that's indeed the case, we say that the
+     term is :gdef:`well-typed`.
```

Comment:
How about presenting the essential info in the following order (if correct):
- every type is a term but not every term is a type
- there is a set of typing rules that can determine the type [a type?] associated with most but not all terms
- terms for which a type is determined are considered well-typed
- Coq generally works with well-typed terms

Mentioning the notation could go in the introduction

Dependent types deserve more explanation (elsewhere) and worth our own definition, not just a Wikipedia link.  Not sure they need to be mentioned here.  "A specificity of" misuses the word.  One could say "In the |Coq| language, types can depend ..."

The footnote didn't seem helpful for the beginner--typing rules and sorts have not been explained yet--and maybe too mundane/out of place for non-beginners.

> The kernel of |Coq| is a type checker that verifies that a term has
     the expected type by applying a set of typing rules (see
     :ref:`Typing-rules`).  If that's indeed the case, we say that the
     term is :gdef:`well-typed`.

I thought the definition is that a term is well-typed if the type can be computed--not that a pair (term, type) is consistent.  It may be true and useful that the Coq kernel checks these pairs but that's not the definition or very relevant right here, right?

Some of this could eventually go into the introduction and with more elaboration.

Reply:
>How about presenting the essential info in the following order (if correct):

I am not convinced by your proposed order. However, I've still moved up the paragraph defining well-typed. Have a look and let me know if there is something specific that you don't like in the new order.

>Dependent types deserve more explanation (elsewhere) and worth our own definition, not just a Wikipedia link. Not sure they need to be mentioned here. "A specificity of" misuses the word. One could say "In the |Coq| language, types can depend ..."

Sure, they can be explained further when we reach the notion of product types. At that point, we can introduce a glossary definition, and then we can link to it from there instead of linking to Wikipedia. But my reason for mentioning dependent types at that point was to justify that types and terms share a common language, whereas this is not usually the case in standard typed programming languages.

>"A specificity of" misuses the word. One could say "In the |Coq| language, types can depend ..."

I'm replacing "A specificity of" by "A special feature of". I didn't like your suggestion because it removed the aspect that this is something that differs from the other languages, which is precisely what I wanted to emphasize.

>The footnote didn't seem helpful for the beginner--typing rules and sorts have not been explained yet--and maybe too mundane/out of place for non-beginners.

I introduced the footnote because I thought you said that the statement "all types are terms, but not all terms are types" is too imprecise. I'm removing it.

---

To answer the rest of your comments:

>most but not all terms

Saying that we can find types for "most" terms does not make any sense. There is a countable number of terms that is divided into a countable number of well-typed terms and a countable number of non-well-typed terms. If we were to use another metric than cardinality (like probability to build a well-typed term), I would intuitively say that well-typed terms are completely outnumbered by non-well-typed-terms.

>Coq generally works with well-typed terms

Generally would be too vague. It should be "virtually always" but then people would wonder what are the exceptions. This is not useful info at this point. That's why I just wrote "To be valid". The only thing that I wanted to express is that there are things that you can construct with the term grammar presented above which won't be valid terms because they won't be well-typed.

>I thought the definition is that a term is well-typed if the type can be computed--not that a pair (term, type) is consistent.

You can have well-typed terms for which the type cannot be computed but if you do provide the type, it can be checked. Type inference is not complete in Coq contrary to a simpler type system like OCaml's. Probably something else to mention somewhere, for example whenever we do get to the point of defining the terms "type inference" and "type checking".

Reply:
Oh, checking is decidable?  Good to know.  Eager to see this info in the doc.

Reply:
> I am not convinced by your proposed order.

It's best to begin with a positive statement about types and then to qualify it.  That's why I'd put "[syntactically] all types are terms" first, after which "but not all terms are types" follows naturally.  Then explain why not all terms are types.

> Saying that we can find types for "most" terms does not make any sense.

Poor wording by me.  Should have been for "some" types.

> Generally would be too vague.

Not at all.  "Generally" means "in general terms; without regard to particulars or exceptions". It could mean 51% of the time or 99.99% of the time.  It implies that there are exceptions but that they're not important enough to mention here.

Reply:
>Oh, checking is decidable? Good to know. Eager to see this info in the doc.

It is, unless you activate some unsafe flags like [`Cumulative StrictProp`](https://coq.github.io/doc/master/refman/addendum/sprop.html#coq:flag.cumulative-strictprop).